include/unifex/filter_stream.hpp

```diff
@@ -0,0 +1,262 @@
+      friend auto tag_invoke(CPO cpo, const type& r) noexcept(
+          is_nothrow_callable_v<CPO, const Receiver&>)
+          -> callable_result_t<CPO, const Receiver&> {
+    return std::move(cpo)(std::as_const(r.op_.receiver_));
+  }
+
+#if UNIFEX_ENABLE_CONTINUATION_VISITATIONS
+  template <typename Func>
+  friend void
+  tag_invoke(tag_t<visit_continuations>, const type& r, Func&& func) {
+    std::invoke(func, r.op_.receiver_);
+  }
+#endif
+
+  template <typename... Values>
+  void set_value(Values&&... values) && noexcept {
+    auto& op = op_;
+    op.next_.destruct();
+    UNIFEX_TRY {
+      const bool doFilter = !std::invoke(op.filter_, std::as_const(values)...);
+
+      if (doFilter) {
+        op.next_.destruct();
+        op.nextEngaged_ = false;
+        op.next_.construct_with([&] {
+          return unifex::connect(next(op.stream_), next_receiver_t{op});
+        });
+        op.nextEngaged_ = true;
+        unifex::start(op.next_.get());
+      } else {
+        unifex::set_value(
+            std::move(op.receiver_), std::forward<Values>(values)...);
+      }
+    }
+    UNIFEX_CATCH(...) {
+      unifex::set_error(std::move(op.receiver_), std::current_exception());
+    }
+  }
+
+  void set_done() && noexcept { unifex::set_done(std::move(op_.receiver_)); }
+
+  template <typename Error>
+  void set_error(Error&& e) && noexcept {
+    unifex::set_error(std::move(op_.receiver_), std::forward<Error>(e));
+  }
+};
+
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+struct _op<StreamSender, FilterFunc, Receiver>::type {
+  StreamSender& stream_;
+  FilterFunc filter_;
+  Receiver receiver_;
+
+  using next_receiver_t = next_receiver<StreamSender, FilterFunc, Receiver>;
+  using next_op =
+      manual_lifetime<next_operation_t<StreamSender, next_receiver_t>>;
+
+  next_op next_;
+  bool nextEngaged_{false};
+
+  template <typename StreamSender2, typename FilterFunc2, typename Receiver2>
+  explicit type(
+      StreamSender2&& stream,
+      FilterFunc2&& filter,
+      Receiver2&&
+          receiver) noexcept(std::
+                                 is_nothrow_constructible_v<
+                                     StreamSender,
+                                     StreamSender2>&&
+                                     std::is_nothrow_constructible_v<
+                                         FilterFunc,
+                                         FilterFunc2>&&
+                                         std::is_nothrow_constructible_v<
+                                             Receiver,
+                                             Receiver2>&&
+                                             is_nothrow_connectable_v<
+                                                 next_sender_t<StreamSender>,
+                                                 Receiver&>)
+    : stream_(std::forward<StreamSender2>(stream))
+    , filter_(std::forward<FilterFunc2>(filter))
```

Comment:
It occurs to me that we've over-complicated this constructor. Sorry for the review noise.

`filter_` can be a reference back out to the value that sits in the stream that owns it, just like `stream_` is. And there's no need to "construct" references from perfectly-forwarded arguments.
```suggestion
  StreamSender& stream_;
  FilterFunc& filter_;
  Receiver receiver_;

  using next_receiver_t = next_receiver<StreamSender, FilterFunc, Receiver>;
  using next_op =
      manual_lifetime<next_operation_t<StreamSender, next_receiver_t>>;

  next_op next_;
  bool nextEngaged_{false};

  template <typename Receiver2>
  explicit type(
      StreamSender& stream,
      FilterFunc& filter,
      Receiver2&&
          receiver) noexcept(std::is_nothrow_constructible_v<
                                             Receiver,
                                             Receiver2>&&
                                             is_nothrow_connectable_v<
                                                 next_sender_t<StreamSender>,
                                                 Receiver&>)
    : stream_(stream)
    , filter_(filter)
```